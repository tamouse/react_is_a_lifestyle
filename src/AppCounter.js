import React from "react"
import PropTypes from "prop-types"
import "./AppCounter.css"

/*

   Discussion
   ==========

   When the currentValue prop is changed in the parent component, we want
   to do something to indicate that it changed. Testing the new property
   value against the previous property value, we can set a state variable
   to "true", and after state has been changed, call a timeout function
   in 500 milliseconds to return that this.state.tate variable to false.

   The display update is handled by adding a classname to the result
   container, just like was done in the jQuery example. It's added
   with a ternary expression that tests the boolean in state, adding
   the class when true, and nothing when false.

   The cool thing here is that all the control for display of the counter
   value, all the interaction provided, is handled within the display
   component, AppCounter, while dealing with the business of keeping
   track of the count and providing the mechanism to change it is handled
   in the Counter component. This lets the Counter functionality be used
   by other counting display components in the app.

 */

export default class AppCounter extends React.Component {
  // STATICS

  static propTypes = {
    currentValue: PropTypes.number.isRequired,
    incrementer: PropTypes.func.isRequired,
    reset: PropTypes.func.isRequired
  }

  // INSTANCE VARIABLES

  state = {
    resultChanged: false
  }

  // INSTANCE METHODS

  handleResultUpdate = () => {
    this.setState({ resultChanged: true }, () => {
      setTimeout(() => {
        this.setState({ resultChanged: false })
      }, 500)
    })
  }

  // LIFECYCLE METHODS

  componentDidUpdate(prevProps, prevState) {
    if (this.props.currentValue !== prevProps.currentValue) {
      this.handleResultUpdate()
    }
  }

  render() {
    const { currentValue, incrementer, reset } = this.props
    return (
      <div className="counter">
        <div className={`result ${this.state.resultChanged ? "changing" : ""}`}>
          {currentValue}
        </div>
        <div>
          <button
            id="dbldecbtn"
            className="counter-button"
            onClick={() => incrementer(-10)}
          >
            --
          </button>{" "}
          <button
            id="decbtn"
            className="counter-button"
            onClick={() => incrementer(-1)}
          >
            -
          </button>{" "}
          <button id="resetbtn" className="counter-button" onClick={reset}>
            C
          </button>{" "}
          <button
            id="incbtn"
            className="counter-button"
            onClick={() => incrementer(+1)}
          >
            +
          </button>{" "}
          <button
            id="dblincbtn"
            className="counter-button"
            onClick={() => incrementer(+10)}
          >
            ++
          </button>
        </div>
      </div>
    )
  }
}
