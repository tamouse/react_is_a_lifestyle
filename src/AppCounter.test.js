import React from "react"
import { shallow } from "enzyme"
import AppCounter from "./AppCounter"

describe("test AppCounter", () => {
  let wrapper, MockIncrementer, MockResetter

  beforeEach(() => {
    MockIncrementer = jest.fn()
    MockResetter = jest.fn()
    wrapper = shallow(
      <AppCounter
        currentValue={0}
        incrementer={MockIncrementer}
        reset={MockResetter}
      />
    )
  })

  it("renders the AppCounter", () => {
    expect(wrapper).toMatchSnapshot()
  })

  it("pushing dbl dec button calls incrementer with -10", () => {
    wrapper.find({ id: "dbldecbtn" }).simulate("click")
    expect(MockIncrementer).toHaveBeenCalled()
    expect(MockIncrementer.mock.calls[0][0]).toBe(-10)
  })

  it("pushing dec button calls incrementer with -1", () => {
    wrapper.find({ id: "decbtn" }).simulate("click")
    expect(MockIncrementer).toHaveBeenCalled()
    expect(MockIncrementer.mock.calls[0][0]).toBe(-1)
  })

  it("pushing inc button calls incrementer with +1", () => {
    wrapper.find({ id: "incbtn" }).simulate("click")
    expect(MockIncrementer).toHaveBeenCalled()
    expect(MockIncrementer.mock.calls[0][0]).toBe(+1)
  })

  it("pushing dblinc button calls incrementer with +10", () => {
    wrapper.find({ id: "dblincbtn" }).simulate("click")
    expect(MockIncrementer).toHaveBeenCalled()
    expect(MockIncrementer.mock.calls[0][0]).toBe(+10)
  })

  it("should show changing class on result update", () => {
    wrapper.setProps({ currentValue: 2 })
    expect(wrapper).toMatchSnapshot()
  })
})
