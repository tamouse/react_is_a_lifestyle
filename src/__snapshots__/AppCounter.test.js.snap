// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test AppCounter renders the AppCounter 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppCounter
    currentValue={0}
    incrementer={[Function]}
    reset={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="result "
/>,
        <div>
          <button
                    className="counter-button"
                    id="dbldecbtn"
                    onClick={[Function]}
          >
                    --
          </button>
           
          <button
                    className="counter-button"
                    id="decbtn"
                    onClick={[Function]}
          >
                    -
          </button>
           
          <button
                    className="counter-button"
                    id="resetbtn"
                    onClick={[Function]}
          >
                    C
          </button>
           
          <button
                    className="counter-button"
                    id="incbtn"
                    onClick={[Function]}
          >
                    +
          </button>
           
          <button
                    className="counter-button"
                    id="dblincbtn"
                    onClick={[Function]}
          >
                    ++
          </button>
</div>,
      ],
      "className": "counter",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 0,
          "className": "result ",
        },
        "ref": null,
        "rendered": 0,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="counter-button"
              id="dbldecbtn"
              onClick={[Function]}
>
              --
</button>,
            " ",
            <button
              className="counter-button"
              id="decbtn"
              onClick={[Function]}
>
              -
</button>,
            " ",
            <button
              className="counter-button"
              id="resetbtn"
              onClick={[Function]}
>
              C
</button>,
            " ",
            <button
              className="counter-button"
              id="incbtn"
              onClick={[Function]}
>
              +
</button>,
            " ",
            <button
              className="counter-button"
              id="dblincbtn"
              onClick={[Function]}
>
              ++
</button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "--",
              "className": "counter-button",
              "id": "dbldecbtn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "--",
            "type": "button",
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "-",
              "className": "counter-button",
              "id": "decbtn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "-",
            "type": "button",
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "C",
              "className": "counter-button",
              "id": "resetbtn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "C",
            "type": "button",
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "+",
              "className": "counter-button",
              "id": "incbtn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "+",
            "type": "button",
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "++",
              "className": "counter-button",
              "id": "dblincbtn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "++",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="result "
/>,
          <div>
            <button
                        className="counter-button"
                        id="dbldecbtn"
                        onClick={[Function]}
            >
                        --
            </button>
             
            <button
                        className="counter-button"
                        id="decbtn"
                        onClick={[Function]}
            >
                        -
            </button>
             
            <button
                        className="counter-button"
                        id="resetbtn"
                        onClick={[Function]}
            >
                        C
            </button>
             
            <button
                        className="counter-button"
                        id="incbtn"
                        onClick={[Function]}
            >
                        +
            </button>
             
            <button
                        className="counter-button"
                        id="dblincbtn"
                        onClick={[Function]}
            >
                        ++
            </button>
</div>,
        ],
        "className": "counter",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 0,
            "className": "result ",
          },
          "ref": null,
          "rendered": 0,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="counter-button"
                id="dbldecbtn"
                onClick={[Function]}
>
                --
</button>,
              " ",
              <button
                className="counter-button"
                id="decbtn"
                onClick={[Function]}
>
                -
</button>,
              " ",
              <button
                className="counter-button"
                id="resetbtn"
                onClick={[Function]}
>
                C
</button>,
              " ",
              <button
                className="counter-button"
                id="incbtn"
                onClick={[Function]}
>
                +
</button>,
              " ",
              <button
                className="counter-button"
                id="dblincbtn"
                onClick={[Function]}
>
                ++
</button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "--",
                "className": "counter-button",
                "id": "dbldecbtn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "--",
              "type": "button",
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "-",
                "className": "counter-button",
                "id": "decbtn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "-",
              "type": "button",
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "C",
                "className": "counter-button",
                "id": "resetbtn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "C",
              "type": "button",
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "+",
                "className": "counter-button",
                "id": "incbtn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "+",
              "type": "button",
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "++",
                "className": "counter-button",
                "id": "dblincbtn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "++",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`test AppCounter should show changing class on result update 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppCounter
    currentValue={2}
    incrementer={[Function]}
    reset={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="result changing"
>
          2
</div>,
        <div>
          <button
                    className="counter-button"
                    id="dbldecbtn"
                    onClick={[Function]}
          >
                    --
          </button>
           
          <button
                    className="counter-button"
                    id="decbtn"
                    onClick={[Function]}
          >
                    -
          </button>
           
          <button
                    className="counter-button"
                    id="resetbtn"
                    onClick={[Function]}
          >
                    C
          </button>
           
          <button
                    className="counter-button"
                    id="incbtn"
                    onClick={[Function]}
          >
                    +
          </button>
           
          <button
                    className="counter-button"
                    id="dblincbtn"
                    onClick={[Function]}
          >
                    ++
          </button>
</div>,
      ],
      "className": "counter",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 2,
          "className": "result changing",
        },
        "ref": null,
        "rendered": 2,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="counter-button"
              id="dbldecbtn"
              onClick={[Function]}
>
              --
</button>,
            " ",
            <button
              className="counter-button"
              id="decbtn"
              onClick={[Function]}
>
              -
</button>,
            " ",
            <button
              className="counter-button"
              id="resetbtn"
              onClick={[Function]}
>
              C
</button>,
            " ",
            <button
              className="counter-button"
              id="incbtn"
              onClick={[Function]}
>
              +
</button>,
            " ",
            <button
              className="counter-button"
              id="dblincbtn"
              onClick={[Function]}
>
              ++
</button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "--",
              "className": "counter-button",
              "id": "dbldecbtn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "--",
            "type": "button",
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "-",
              "className": "counter-button",
              "id": "decbtn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "-",
            "type": "button",
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "C",
              "className": "counter-button",
              "id": "resetbtn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "C",
            "type": "button",
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "+",
              "className": "counter-button",
              "id": "incbtn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "+",
            "type": "button",
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "++",
              "className": "counter-button",
              "id": "dblincbtn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "++",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="result changing"
>
            2
</div>,
          <div>
            <button
                        className="counter-button"
                        id="dbldecbtn"
                        onClick={[Function]}
            >
                        --
            </button>
             
            <button
                        className="counter-button"
                        id="decbtn"
                        onClick={[Function]}
            >
                        -
            </button>
             
            <button
                        className="counter-button"
                        id="resetbtn"
                        onClick={[Function]}
            >
                        C
            </button>
             
            <button
                        className="counter-button"
                        id="incbtn"
                        onClick={[Function]}
            >
                        +
            </button>
             
            <button
                        className="counter-button"
                        id="dblincbtn"
                        onClick={[Function]}
            >
                        ++
            </button>
</div>,
        ],
        "className": "counter",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 2,
            "className": "result changing",
          },
          "ref": null,
          "rendered": 2,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="counter-button"
                id="dbldecbtn"
                onClick={[Function]}
>
                --
</button>,
              " ",
              <button
                className="counter-button"
                id="decbtn"
                onClick={[Function]}
>
                -
</button>,
              " ",
              <button
                className="counter-button"
                id="resetbtn"
                onClick={[Function]}
>
                C
</button>,
              " ",
              <button
                className="counter-button"
                id="incbtn"
                onClick={[Function]}
>
                +
</button>,
              " ",
              <button
                className="counter-button"
                id="dblincbtn"
                onClick={[Function]}
>
                ++
</button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "--",
                "className": "counter-button",
                "id": "dbldecbtn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "--",
              "type": "button",
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "-",
                "className": "counter-button",
                "id": "decbtn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "-",
              "type": "button",
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "C",
                "className": "counter-button",
                "id": "resetbtn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "C",
              "type": "button",
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "+",
                "className": "counter-button",
                "id": "incbtn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "+",
              "type": "button",
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "++",
                "className": "counter-button",
                "id": "dblincbtn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "++",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
